File: reader.rb
#!/usr/bin/ruby

require 'rubygems'
require 'large.pb'
require 'small.pb'

SMALL = true
LARGE = false

SMALL_T = 5
LARGE_T = 50

bufclass = (ARGV[0] == "small")

fields = 0

case bufclass
when SMALL
  data = Small::Data.new
when LARGE
  data = Large::Data.new
else
  raise "Unknown error"
end
counter = 0
while ((s = STDIN.readline) rescue false)
  s = STDIN.readline
  s.gsub!("\n", "")
  s.gsub!("\r", "")
  data.parse_from_string(s)
  counter += 1
  if counter % 10000 == 0 
    warn "READER: #{counter} records passed..."
  end
end

File: reader.cpp
#include <iostream>
#include <fstream>
#include <string>
#include "large.pb.cc"
#include "small.pb.cc"

using namespace std;

void small_data(small::Data *data)
{
  unsigned int counter = 0;
  string str;
  while (true) {
    getline(cin, str);
    getline(cin, str);
    data->ParseFromString(str);
    counter++;
    if(counter % 10000 == 0){
      cerr << "READER: " << counter << " records passed" << endl;
    }
  }
}

void large_data(large::Data *data)
{
  unsigned int counter = 0;
  string str;
  while (true) {
    getline(cin, str);
    getline(cin, str);
    !data->ParseFromString(str);
    counter++;
    if(counter % 10000 == 0){
      cerr << "READER: " << counter << " records passed" << endl;
    }
  }
}

int main(int argc, char* argv[]) {

  GOOGLE_PROTOBUF_VERIFY_VERSION;

  if (argc < 2) {
    cerr << "Usage:  " << argv[0] << " PROTO_CLASS" << endl;
    return -1;
  }
 
  string bufclass = argv[1];

  if (bufclass == "small"){
    small::Data data;
    small_data(&data);
  } else {
    large::Data data;
    large_data(&data);
  } 

  google::protobuf::ShutdownProtobufLibrary();

  return 0;
}
